In [1]:
import pandas as pd
df = pd.read_csv("usarrests.csv")

In [2]:
df. head()

,Unnamed: 0,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91


In [3]:
df.shape

(50, 4)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  50 non-null     object 
 1   Murder      50 non-null     float64
 2   Assault     49 non-null     float64
 3   UrbanPop    50 non-null     int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 1.7+ KB


In [5]:
df.describe()

,Murder,Assault,UrbanPop
count,50.00000,49.000000,50.00000
mean,7.78800,182.183673,74.20000
std,4.35551,130.877435,73.40828
min,0.80000,45.000000,6.00000
25%,4.07500,109.000000,53.25000
50%,7.25000,159.000000,66.00000
75%,11.25000,249.000000,77.75000
max,17.40000,879.000000,570.00000


## Initial Observations
    1. The count for assault is 49 but there is 50 rows so there is a missing value. 
    2. The Urban Population has a max percentage of of 570% , which seems suspicious since it is out of 100%.
    3. The min percentage of 6 for UrbanPop seems suspicious in comparison to the other values.
    4. The max assualt value is 879, which seems unusally high in comparison with the mean of 182.18 and median 159.

In [6]:
df.isna().sum()

Unnamed: 0    0
Murder        0
Assault       1
UrbanPop      0
dtype: int64

In [7]:
fill_value = df["Assault"].median()
df["Assault"] = df["Assault"].fillna(fill_value)

In [8]:
print (fill_value)

159.0


In [9]:
df.isna().sum()

Unnamed: 0    0
Murder        0
Assault       0
UrbanPop      0
dtype: int64

In [10]:
 print(df.duplicated().sum())

0


In [11]:
df[df["UrbanPop"] >100]

,Unnamed: 0,Murder,Assault,UrbanPop
14,Iowa,2.2,56.0,570


In [12]:
df.loc[14,"UrbanPop"] = 57

In [13]:
df[df["UrbanPop"] >100]

,Unnamed: 0,Murder,Assault,UrbanPop


##  Checkpoint
Checking new mean, min, max and median after correcting UrbanPop 570 value

In [14]:
df.describe()

,Murder,Assault,UrbanPop
count,50.00000,50.000000,50.000000
mean,7.78800,181.720000,63.940000
std,4.35551,129.576554,16.453286
min,0.80000,45.000000,6.000000
25%,4.07500,109.000000,53.250000
50%,7.25000,159.000000,66.000000
75%,11.25000,249.000000,76.500000
max,17.40000,879.000000,91.000000


In [15]:
df[df["UrbanPop"] <50]

,Unnamed: 0,Murder,Assault,UrbanPop
1,Alaska,10.0,263.0,48
23,Mississippi,16.1,259.0,44
31,New York,11.1,254.0,6
32,North Carolina,13.0,337.0,45
33,North Dakota,0.8,45.0,44
39,South Carolina,14.4,879.0,48
40,South Dakota,3.8,86.0,45
44,Vermont,2.2,48.0,32
47,West Virginia,5.7,81.0,39


In [16]:
df[df["Assault"] >250]

,Unnamed: 0,Murder,Assault,UrbanPop
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
4,California,9.0,276.0,91
8,Florida,15.4,335.0,80
19,Maryland,11.3,300.0,67
21,Michigan,12.1,255.0,74
23,Mississippi,16.1,259.0,44
27,Nevada,12.2,252.0,81
30,New Mexico,11.4,285.0,70
31,New York,11.1,254.0,6


### Flagged Values
    - UrbanPop: 570%- Impossible 
    - Assault: 879- Extreme
    - UrbanPop: 6%- Extreme

## Cleaning Log
    - Filled 1 missing assault values with the median of median 159. The maximum assault is 879, the mean is 182.18 and median is 159. I chose the median because the very high maximum can skew the data and mean.
    - 570% is impossible for UrbanPop since it is a percentage and a percentage can't exceed 100%. I assumed an extra 0 was added when entering the data. Based on  this assumption I replaced the value with 57%, which fits in the range of the data set with the mean of 74.2 and median of 66. After the correcting the vale, the UrbanPop mean decreased from 74.20 to 63.94,the median remained 66. This shows how the 570 was skewing the mean.  
    - I think the maximum assault value of 879 is an extreme and not an incorrect value. Since assault is measured per 100,000 residents 879 is possible. Although this value is unusally high, I decided to keep the value unchanged since I don't have enough evidence to determine that it's an error.
    - I think the unsually low UrbanPop value of 6% is an extreme and not an incorrect value. This value is possble since it is a  percentage that ranges from 0-100%. Thus, I decided to keep the value unchanged since I don't have enough evidence to detemine that it's an error. 

Git Hub Link: https://github.com/jazzminowens/cs82a-portfolio